src/Language/Futhark/TypeChecker/Terms.hs

```diff
@@ -439,7 +662,8 @@ checkExp (AppExp (Apply fe args loc) NoInfo) = do
           Var v _ _ -> Just v
           _ -> Nothing
   ((_, exts, rt), args'') <- mapAccumLM (onArg fname) (0, [], t) args'
-  pure $ AppExp (Apply fe' args'' loc) $ Info $ AppRes rt exts
+  (rt', exts') <- unscopeUnknowable loc rt
```

Comment:
Similarly, apparently also unnecessary.

Reply:
It's for the rule for not allowing existential size that are not in scope to appear in size expressions.

I've made this for `tests/shapes/existential-argument.fut`, and more specifically for a case like
```futhark
def quad_eval 't (f : () -> []t) : []t =
    let t = f () ++ f ()
    in t ++ t
```
Which is monomorphised as
```futhark
def double_eval ((f: () -> ?[d₀].[d₀]i64)) : ?[d₁₀].[d₁₀]i64 =
  let [d<{(+) d₅ d₆}>] [d₅] [d₆] (t: [d<{(+) d₅ d₆}>]i64) =
    (++) (f ()) (f ())
  let (d<{(+) ((+) d₅ d₆) ((+) d₅ d₆)}>: i64) =
    (+) ((+) d₅ d₆) ((+) d₅ d₆)
  in (++) t t
```
Somehow these `d₅` and `d₆` does not crash the compiler in latter stages, but I'm more confident with this.

Reply:
Interesting!  I removed this code and now this function

```
def double_eval 't (f : () -> []t) : []t =
    f () ++ f ()
```

is type-checked as

```Futhark
def double_eval_6007 't_5993 ((f_5995: () -> ?[d₀_5994].[d₀_5994]t_5993): () -> []t_5993) : [d₅_6001 +_494 d₆_6003]t_5993 =
  f_5995 () ++_5618 f_5995 ()
```

It is clearly nonsensical to refer to free variables `d₅_6001` and `d₆_6003` in the return type.  However, the fix should not be to strip unknown sizes in application - those are fine *locally*.  But we should strip them when generalising the type of functions.

Interestingly, we do this correctly for functions such as

```Futhark
def double_eval 't (f : () -> []t) : []t =
    f ()
```

which get type-checked as

```Futhark
def double_eval_6000 't_5993 ((f_5995: () -> ?[d₀_5994].[d₀_5994]t_5993): () -> []t_5993) : ?[d₂_5998].[d₂_5998]t_5993 =
  f_5995 ()
```

Reply:
Note the explicit quantification of the existential `d₂_5998` in the return type of the last example above.  That's the thing we're missing.  Probably the implementation of generalisation only considers size expressions that are plain variables.

Reply:
yes but
```futhark
def double_eval n (f : () -> []i64) : []i64 =
    f () ++ iota n
```
is type-checked as
```futhark
def double_eval (n: i64) ((f: () -> ?[d₁].[d₁]i64): () -> []i64) : [d₆ + n]i64 =
  f () ++ iota n
```
which is an incorrect type

Reply:
and allow the following to type-check!
```futhark
def main (n:i64) =
    zip (double_eval n (\() -> iota n)) (double_eval n (\() -> iota 2))
```

Reply:
I think the culprit is `noSizeEscape` in `checkFunBody` which removes these unknown sizes.

Reply:
maybe just call it over all return type ? I'm not sure they are evaluated

Reply:
No, wait, that's not the case.  `noSizeEscape` is actually quite innocent.

Reply:
we should also free them in function arguments, or correct monomorphisation regarding these sizes:
```futhark
def foo [n][m]'t (_ : [n+m]t) : (i64,i64) =
  (n, m)

def double_eval n (f : () -> []i64) =
    foo (f () ++ iota n)
```
monomorphised as
```futhark
def foo [d<{(+) n m}>] (n: i64) (m: i64) ((_: [d<{(+) n m}>]i64)) : (i64, i64) =
  (n, m)

def double_eval (n: i64) ((f: () -> ?[d₁].[d₁]i64)) : (i64, i64) =
  foo 0i64 0i64 ((++) (f ()) (iota n))
```
which is obviously wrong.

If not correcting monomorphisation, don't we end up with same expressivity ?

Reply:
The problem is in `closeOverTypes`, specifically the way it uses `injectExt` to figure out where to put the existential.  The current logic is to put it at the first constructive use - but there might not *be* a constructive use!  In this case we should replace the entire dimension with a fresh unknown size, just like what `unscopeUnknowable` used to do (but it should be done at function level, not after any application).

Reply:
To motivate, the reason I do not like invoking `unscopeUnknowable` after every application is because it disallowed this:

```Futhark
def double_eval 't (f : () -> []t) =
  let x = f ()
  let y = f ()
  in zip (x ++ y) (x ++ y)
```

Reply:
no, sizes of x and y will be marked as bounded by the let-bindings

Reply:
that's what allow us to write
```futhark
  let x = filter ...
  in map (x++) ...
```

Reply:
Oh, I had not noticed that machinery.  Is *that* still necessary if we switch to only using `unscopeUnknowable` (or equivalent) on function bodies?

Reply:
I get that this is more similar to the paper, but I'm not sure of the gain - and I am uncomfortable with having size variables that are also bound in the environment.

Reply:
I have not pondered the interaction with monomorphisation - I'm still stuck in a type checking mindset.

Reply:
I don't think so, we have to check what happen for this program
```futhark
def double_eval [n] (f : () -> []i64) (ma:[][n]i64) =
    let t = f()
    in map (t ++) ma
```
as
```futhark
def double_eval [n] (f : () -> []i64) (ma:[][n]i64) =
    map (f () ++) ma
```
still rejected, I've done the scoping machinery to change the nature of the size variable introduced by f

Reply:
removing let machinery lead to refuse
```futhark
def double_eval [n] (f : () -> []i64) (ma:[][n]i64) =
    let t = f()
    in map (t ++) ma
```

Reply:
> I am uncomfortable with having size variables that are also bound in the environment.

Why ? we can't refers to them, and they are freed the same way as the variable introduced in a 'classic' way, I don't think they can harm

Reply:
I also tried removing the `let` machinery, and the program you posted above (with the binding of `t`) type checks fine for me.  The one without first binding `t = f()` fails with a causality violation, which matches old (and safe) behaviour.

Reply:
This is what I did:

```
diff --git a/src/Language/Futhark/TypeChecker/Terms.hs b/src/Language/Futhark/TypeChecker/Terms.hs
index e2c66cd9c..5a9a6a8e5 100644
--- a/src/Language/Futhark/TypeChecker/Terms.hs
+++ b/src/Language/Futhark/TypeChecker/Terms.hs
@@ -671,15 +671,11 @@ checkExp (AppExp (LetPat sizes pat e body loc) _) =
     constraints <- getConstraints
     incLevel . bindingSizes sizes $ \sizes' ->
       bindingPat sizes' pat (Ascribed t) $ \pat' -> do
-        let implicitSizes = filter (isUnknown constraints) $ M.keys $ unFV $ freeInPat pat'
-            implicitIdents = map (\vn -> Ident vn (Info $ Scalar $ Prim $ Signed Int64) mempty) implicitSizes
-        binding False implicitIdents $ do
-          mapM_ observe implicitIdents
-          body' <- checkExp body
-          (body_t, retext) <-
-            unscopePatType loc (sizesMap sizes' <> patNames pat' <> S.fromList implicitSizes) =<< expTypeFully body'
-
-          pure $ AppExp (LetPat sizes' pat' e' body' loc) (Info $ AppRes body_t retext)
+        body' <- checkExp body
+        (body_t, retext) <-
+          unscopePatType loc (sizesMap sizes' <> patNames pat') =<< expTypeFully body'
+
+        pure $ AppExp (LetPat sizes' pat' e' body' loc) (Info $ AppRes body_t retext)
   where
     sizesMap = foldMap (S.singleton . sizeName)
     isUnknown constraints vn
```

Reply:
I had done the same thing, I think that I had not fully remove unscopeUnknowable.

Then I don't think we still need it, it's also done in Match.

Reply:
`tests/operator/size-section2.fut` (tricky program!) monomorphises to an invalid program with these changes, but it might just depend on a subtle implementation detail that's no longer true.  I will investigate.

Reply:
Oh, it's because I forgot to actually bring back `unscopeKnowable` for function return types...

Reply:
Nope, that was indeed something I had forgotten, but it was unrelated.